### PR TITLE
DOC, ENH: Support memory_map for Python engine

### PR DESCRIPTION
**Pull request by gfyoung**
Title is self-explanatory.

### CODE REVIEW

File: doc/source/io.rst [hunk 1]
```diff
@@ -193,6 +193,10 @@ use_unsigned : boolean, default False
 
   If integer columns are being compacted (i.e. ``compact_ints=True``), specify whether
   the column should be compacted to the smallest signed or unsigned integer dtype.
+memory_map : boolean, default False
```
Comment (jreback) [inline doc/source/io.rst]: is doc-string updated?

File: pandas/io/tests/parser/c_parser_only.py [hunk 2]
```diff
@@ -307,10 +307,6 @@ def test_usecols_dtypes(self):
         self.assertTrue((result.dtypes == [object, np.int, np.float]).all())
         self.assertTrue((result2.dtypes == [object, np.float]).all())
 
-    def test_memory_map(self):
-        # it works!
```
Comment (jreback) [inline pandas/io/tests/parser/c_parser_only.py]: are you actually changing machinrery for c-engine?

Comment (gfyoung) [inline pandas/io/tests/parser/c_parser_only.py]: No change for C engine.  Just added stuff for Python.

File: doc/source/io.rst (hunk 1 above)
Comment (gfyoung) [inline doc/source/io.rst]: Yes?  I just added it.

File: pandas/io/common.py [hunk 3]
```diff
@@ -324,9 +325,51 @@ def _get_handle(path, mode, encoding=None, compression=None):
         else:
             f = open(path, mode)
 
+    if memory_map and hasattr(f, 'fileno'):
+        try:
+            f = MMapWrapper(f)
+        except:  # fallback: leave file handler as is
```
Comment (shoyer) [inline pandas/io/common.py]: Can you make this more specific? Which exceptions do you need to catch here?

**Comment by shoyer** [comment]
Do you have any benchmarks showing positive performance gains?

If this is an unambiguous win, we should enable it by default rather than adding yet another option.

**Comment by gfyoung** [comment]
I can run benchmarks but remember this is AFAICT a performance vs memory tradeoff given what memory_map does.  I am inclined to leave this still as an option, though how many people use it I am not sure.

File: pandas/io/common.py (hunk 3 above)
Comment (gfyoung) [inline pandas/io/common.py]: It's really whatever can be thrown by the mmap constructor and while I could certainly guess what some of them might be, not really sure how safe that is.

**Comment by gfyoung** [comment]
Ran basic `%timeit` with CSV file sizes of 10000 rows and 1000000 rows, and the performance flip-flops between the two.  I think the I/O overhead starts to count as the file sizes get larger and larger, but I don't know if there is a specific cutoff.  I think this would indicate leaving it as an option until there are more conclusive results.

**Comment by jreback** [comment]
@shoyer this is an *already* existing option, @gfyoung is just documenting.

File: pandas/io/common.py [hunk 4]
```diff
@@ -324,9 +325,51 @@ def _get_handle(path, mode, encoding=None, compression=None):
...
+
+    """
+
+    def __init__(self, f):
+        self.mmap = mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ)
+
+    def __getattr__(self, name):
+        return getattr(self.mmap, name)
+
+    def __next__(self):
```
Comment (jreback) [inline pandas/io/common.py]: if you are going to add this, then pls setup some tests for it in ``io.common/tests/test_common.py``

File: pandas/io/common.py (hunk 3 above)
Comment (jreback) [inline pandas/io/common.py]: I agree with @shoyer here, I think you shouldn't be catching things. Let the exception bubble up; this is a low-level handler and shouldn't hide things.

Comment (jreback) [inline pandas/io/common.py]: tests would help narrow this down! e.g. first test should be a non-existant file, 2nd a file-like which cannot be memory mapped, e.g. StringIO (I think).

**Comment by gfyoung** [comment]
I would also add that this PR does provide support for `memory_map` in the Python engine via Python's `mmap` library.

File: pandas/io/common.py (hunk 3 above)
Comment (gfyoung) [inline pandas/io/common.py]: That's inconsistent with what's done on the C engine side, where if the lower-level function calls fail, we just return `NULL` for the source as you can see <a href="https://github.com/pydata/pandas/blob/master/pandas/parser.pyx#L612">here<a>.

Comment (gfyoung) [inline pandas/io/common.py]: Also, non-existent files won't show up when you instantiate `MMapWrapper`, as the constructor is expecting a `file` object, which should already be pointing to an existing or newly-created file.

Comment (jreback) [inline pandas/io/common.py]: oh, ok, then be explicit about this in the docs for the wrapper. (and test as much as possible).

Comment (gfyoung) [inline pandas/io/common.py]: If you're referring to the `file` object, I think I am already (I wrote a parameters section for that reason).  Or are you referring to the catch-all `try-except` block?

File: pandas/io/common.py (hunk 4 above)
Comment (gfyoung) [inline pandas/io/common.py]: Tests added

File: pandas/io/common.py (hunk 3 above)
Comment (shoyer) [inline pandas/io/common.py]: @gfyoung If you scroll down in `parser.pyx` [a little further](https://github.com/pydata/pandas/blob/master/pandas/parser.pyx#L626), you notice if the pointer is `NULL` we raise `IOError`.

I don't think we should ignore arbitrary errors, certainly not if this is disabled by default. Even then, never use blanket `except:` clauses -- `except Exception:` is much safer.

**Comment by shoyer** [comment]
Just to clarify, this is already supported by the C engine via `**kwargs`, but is currently undocumented?

I'm OK with documenting it and adding it as another argument then -- we already have quite a few arguments for `read_csv`, adding another one won't make things much worse.

**Comment by gfyoung** [comment]
@shoyer : If you look at the signature <a href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html">here</a>, you will see that it is already in the signature **explicitly**.  However, no explanation of it is given.  Also, read @jreback comment <a href="https://github.com/pydata/pandas/pull/13381#issuecomment-224262023">above</a>.

File: pandas/io/common.py (hunk 3 above)
Comment (gfyoung) [inline pandas/io/common.py]: @shoyer : that's not the right `NULL` - this is the one I'm referring to <a href="https://github.com/pydata/pandas/blob/master/pandas/parser.pyx#L613">here</a>.  The one you're pointing at is when all other attempts to create a source have failed.

Comment (shoyer) [inline pandas/io/common.py]: OK, I guess you're right. Still, let's only catch `Exception` and make a note about why we do this.

I'm not sure that we need to consider the current "catch anything" behavior as API but I guess it's better safe than sorry.

Comment (gfyoung) [inline pandas/io/common.py]: However, I have no issues with adding `except Exception`, so I'll do that.

**Comment by gfyoung** [comment]
Travis was having build issues, so I'm rerunning tests.  @jreback could you cancel this old build <a href="https://travis-ci.org/pydata/pandas/builds/135889496">here</a>?

**Comment by gfyoung** [comment]
@jreback : Added the `except Exception` block as @shoyer requested, and Travis is giving the green light.  Ready to merge if there are no other concerns.

**Comment by gfyoung** [comment]
@jreback : Added the `except Exception` block as @shoyer requested, and Travis is giving the green light.  Ready to merge if there are no other concerns.

File: pandas/io/parsers.py [hunk 5]
```diff
@@ -261,7 +261,10 @@
     If integer columns are being compacted (i.e. `compact_ints=True`), specify
     whether the column should be compacted to the smallest signed or unsigned
     integer dtype.
-
+memory_map : boolean, default False
+    If a filepath is provided for `filepath_or_buffer`, map the file object
+    directly onto memory and access the data directly from there. Using this
+    option can improve performance because there is no longer any I/O overhead.
 Returns
```
Comment (jorisvandenbossche) [inline pandas/io/parsers.py]: Blank line is missing before 'Returns'

Comment (gfyoung) [inline pandas/io/parsers.py]: Done.

**Comment by gfyoung** [comment]
@jreback : Added the `except Exception` block as @shoyer requested, and Travis is giving the green light.  Ready to merge if there are no other concerns.

**Comment by jreback** [comment]
thanks!

**Pull request closed without merging**